src/common/base-store.ts

```diff
@@ -0,0 +1,142 @@
+import path from "path"
+import Config from "conf"
+import { Options as ConfOptions } from "conf/dist/source/types"
+import { app, ipcMain, IpcMainEvent, ipcRenderer, IpcRendererEvent, remote } from "electron"
+import { action, observable, reaction, runInAction, toJS, when } from "mobx";
+import Singleton from "./utils/singleton";
+import { getAppVersion } from "./utils/app-version";
+import logger from "../main/logger";
+import { broadcastIpc } from "./ipc";
+import isEqual from "lodash/isEqual";
+
+export interface BaseStoreParams<T = any> extends ConfOptions<T> {
+  autoLoad?: boolean;
+  syncEnabled?: boolean;
+}
+
+export class BaseStore<T = any> extends Singleton {
+  protected storeConfig: Config<T>;
+  protected syncDisposers: Function[] = [];
+
+  whenLoaded = when(() => this.isLoaded);
+  @observable isLoaded = false;
+  @observable protected data: T;
+
+  protected constructor(protected params: BaseStoreParams) {
+    super();
+    this.params = {
+      autoLoad: false,
+      syncEnabled: true,
+      ...params,
+    }
+    this.init();
```

Comment:
I think this call is fundamentally broken. It is not awaited (so when the `getInstance` call finishes the store isn't even up to date) and the default params to init block from enabling sync.

Reply:
Wat? :) can you elablorate/provide more examples?
When you call `getInstance()` you always have an instance of the store. 
Yes, it might be not loaded with a data yet because of `async load()` and it doesn't mean it is ready-to-use immediately. Every store has `isLoaded` field which takes in account before app ready to start.

Reply:
I mean that it seems like a massive foot gun to get one of these and not be able to get the correct data immediately afterwards. 

What I mean is that even if you pass into `getInstance` the argument `{ parms: { autoLoad: true } }` it won't be correct.

Further more, in the default case it seems weird for the unbounded promise (which in future versions of node will abort the process if it throws) halts on `await this.whenLoaded`. I understand that the point of this is so that `enableSync` can be started only once the store has loaded but the current method seems clunky.

Reply:
> not be able to get the correct data immediately afterwards

how you can get it immediately for async store loading? (yes, currently it can be loaded sync from file-system, but it was developed to support async load)

didn't get what you mean about `await this.whenLoaded` and what's wrong with it?
electron has similar stuff, like `app.whenReady`